### ISSUE TITLE
retract/1 bug apparently when retracting rules

### ISSUE THREAD

**Issue by pmoura** [infradig/trealla#196]
I'm observing a `retract/1` bug apparently when retracting rules (after a trace). One example is when running the tests for the Logtalk `gensym` library:

```text
!     gensym_gensym_2_01: failure (in 0.0002727508544921875 seconds)
!       test goal throws an error but should have succeeded: error(instantiation_error,not_sufficiently_instantiated)
!       in file /Users/pmoura/logtalk/library/gensym/tests.lgt between lines 0-4
!     gensym_gensym_2_02: failure (in 5.1021575927734375e-05 seconds)
!       test goal throws an error but should have succeeded: error(instantiation_error,not_sufficiently_instantiated)
!       in file /Users/pmoura/logtalk/library/gensym/tests.lgt between lines 4-2
!     gensym_reset_gensym_0_01: failure (in 5.1021575927734375e-05 seconds)
!       test goal throws an error but should have succeeded: error(instantiation_error,not_sufficiently_instantiated)
!       in file /Users/pmoura/logtalk/library/gensym/tests.lgt at or above line 2
!     gensym_reset_gensym_0_02: failure (in 4.8160552978515625e-05 seconds)
!       test goal throws an error but should have succeeded: error(instantiation_error,not_sufficiently_instantiated)
!       in file /Users/pmoura/logtalk/library/gensym/tests.lgt between lines 2-3
!     gensym_reset_gensym_1_02: failure (in 4.6968460083007812e-05 seconds)
!       test goal throws an error but should have succeeded: error(instantiation_error,not_sufficiently_instantiated)
!       in file /Users/pmoura/logtalk/library/gensym/tests.lgt between lines 2-3
!     gensym_reset_gensym_1_03: failure (in 4.7206878662109375e-05 seconds)
!       test goal throws an error but should have succeeded: error(instantiation_error,not_sufficiently_instantiated)
!       in file /Users/pmoura/logtalk/library/gensym/tests.lgt at or above line 3
```

The library is compiled in debug mode (in order to collect code coverage stats), which result in facts instead of rules being retracted. Maybe is possible for you to debug this issue on the C side by instrumenting the code? This bug also causes test failures in examples such as `aspects`.

**Comment by infradig** [comment]
If I pull the gensym code to a stand-alone .pl file it works perfectly....

**Comment by infradig** [comment]
This is not fixed by the recent fix , just to be clear.

**Comment by pmoura** [comment]
> If I pull the gensym code to a stand-alone .pl file it works perfectly....

If you do that, the code retracts **facts**. As I tried to explain, the compilation of the code for collecting cove coverage stats results in the code retracting **rules** and exposes a bug in `retract/1`.

**Comment by infradig** [comment]
It's fixed and runs.

**Comment by pmoura** [comment]
Confirmed that the second fix (542ca8ed80ef1e530e9fe14dc5056ef0153f25c3) fixes the test failures in `gensym`, `aspects`, `viewpoints`, and `retract/1` tests 👍
